optaplanner-core/src/main/java/org/optaplanner/core/impl/partitionedsearch/queue/PartitionQueue.java

```diff
@@ -0,0 +1,159 @@
+
+    // Only used by consumer
+    private int openPartCount;
+    private final Map<Integer, Long> processedEventIndexMap; // Key is partIndex
+
+    public PartitionQueue(int partCount) {
+        // TODO partCount * 100 is pulled from thin air
+        queue = new ArrayBlockingQueue<>(partCount * 100);
+        moveEventMap = new ConcurrentHashMap<>(partCount);
+        Map<Integer, AtomicLong> nextEventIndexMap = new HashMap<>(partCount);
+        for (int i = 0; i < partCount; i++) {
+            nextEventIndexMap.put(i, new AtomicLong(0));
+        }
+        this.nextEventIndexMap = Collections.unmodifiableMap(nextEventIndexMap);
+        openPartCount = partCount;
+        // HashMap because only the consumer thread uses it
+        processedEventIndexMap = new HashMap<>(partCount);
+        for (int i = 0; i < partCount; i++) {
+            processedEventIndexMap.put(i, -1L);
+        }
+    }
+
+    /**
+     * Thread-safe method.
+     * The previous move(s) for this partIndex (if it hasn't been consumed yet), will be skipped during iteration.
+     * @param partIndex {@code 0 <= partIndex < partCount}
+     * @param move never null
+     */
+    public void addMove(int partIndex, PartitionChangeMove<Solution_> move) {
+        long eventIndex = nextEventIndexMap.get(partIndex).getAndIncrement();
+        PartitionChangedEvent<Solution_> event = new PartitionChangedEvent<>(
+                partIndex, eventIndex, move);
+        moveEventMap.put(event.getPartIndex(), event);
+        queue.add(event);
+    }
+
+    /**
+     * Thread-safe method.
+     * The previous move for this partIndex (if it hasn't been consumed yet), will still be returned during iteration.
+     * @param partIndex {@code 0 <= partIndex < partCount}
+     */
+    public void addFinish(int partIndex) {
+        long eventIndex = nextEventIndexMap.get(partIndex).getAndIncrement();
+        PartitionChangedEvent<Solution_> event = new PartitionChangedEvent<>(
+                partIndex, eventIndex, PartitionChangedEvent.PartitionChangedEventType.FINISHED);
+        queue.add(event);
+    }
+
+    /**
+     * Thread-safe method.
+     * The previous move for this partIndex (if it hasn't been consumed yet), will still be returned during iteration
+     * before the iteration throws an exception.
+     * @param partIndex {@code 0 <= partIndex < partCount}
+     * @param throwable never null
+     */
+    public void addExceptionThrown(int partIndex, Throwable throwable) {
+        long eventIndex = nextEventIndexMap.get(partIndex).getAndIncrement();
+        PartitionChangedEvent<Solution_> event = new PartitionChangedEvent<>(
+                partIndex, eventIndex, throwable);
+        queue.add(event);
+    }
+
+    @Override
+    public Iterator<PartitionChangeMove> iterator() {
+        // Currently doesn't be support to be called twice on the same instance
+        return new PartitionQueueIterator();
```

Comment:
Currently it will never be true, because the phase uses it once and throws it away. You're right though, should be defensive. Not sure what to do yet... maybe just add a todo.